polkadot/node/network/availability-recovery/src/task.rs

```diff
@@ -0,0 +1,851 @@
+	) -> Self {
+		Self { sender, params, strategies, state: State::new() }
+	}
+
+	async fn in_availability_store(&mut self) -> Option<AvailableData> {
+		if !self.params.bypass_availability_store {
+			let (tx, rx) = oneshot::channel();
+			self.sender
+				.send_message(AvailabilityStoreMessage::QueryAvailableData(
+					self.params.candidate_hash,
+					tx,
+				))
+				.await;
+
+			match rx.await {
+				Ok(Some(data)) => return Some(data),
+				Ok(None) => {},
+				Err(oneshot::Canceled) => {
+					gum::warn!(
+						target: LOG_TARGET,
+						candidate_hash = ?self.params.candidate_hash,
+						"Failed to reach the availability store",
+					)
+				},
+			}
+		}
+
+		None
+	}
+
+	/// Run this recovery task to completion. It will loop through the configured strategies
+	/// in-order and return whenever the first one recovers the full `AvailableData`.
+	pub async fn run(mut self) -> Result<AvailableData, RecoveryError> {
+		if let Some(data) = self.in_availability_store().await {
+			return Ok(data)
+		}
+
+		self.params.metrics.on_recovery_started();
+
+		let _timer = self.params.metrics.time_full_recovery();
+
+		let res = loop {
+			if let Some(mut current_strategy) = self.strategies.pop_front() {
+				// Make sure we are not referencing futures from past RecoveryStrategy runs.
+				if self.state.requesting_chunks.total_len() != 0 {
+					self.state.requesting_chunks = FuturesUndead::new();
+				}
+
+				gum::info!(
+					target: LOG_TARGET,
+					candidate_hash = ?self.params.candidate_hash,
+					"Starting `{}` strategy",
+					current_strategy.display_name(),
+				);
+
+				let res =
+					current_strategy.run(&mut self.state, &mut self.sender, &self.params).await;
+
+				match res {
+					Err(RecoveryError::Unavailable) =>
+						if self.strategies.front().is_some() {
+							gum::warn!(
+								target: LOG_TARGET,
+								candidate_hash = ?self.params.candidate_hash,
+								"Recovery strategy `{}` did not conclude. Trying the next one.",
+								current_strategy.display_name(),
+							);
+							continue
+						},
+					Err(err) => break Err(err),
+					Ok(data) => break Ok(data),
+				}
+			} else {
+				// We have no other strategies to try.
+				gum::error!(
+					target: LOG_TARGET,
+					candidate_hash = ?self.params.candidate_hash,
+					"Recovery of available data failed.",
+				);
+				break Err(RecoveryError::Unavailable)
+			}
+		};
+
+		match &res {
+			Ok(_) => self.params.metrics.on_recovery_succeeded(),
+			Err(RecoveryError::Invalid) => self.params.metrics.on_recovery_invalid(),
+			Err(_) => self.params.metrics.on_recovery_failed(),
+		}
+
+		res
+	}
+}
+
+/// `RecoveryStrategy` that sequentially tries to fetch the full `AvailableData` from
+/// already-connected validators in the configured validator set.
+pub struct FetchFull {
+	params: FetchFullParams,
+}
+
+pub struct FetchFullParams {
+	/// Name of the validator group used for recovery. For logging purposes.
```

Comment:
Approval checkers are not really a group. In general all candidates are assigned to all validators for checking across many tranches.

I don't think we really need this string here. The strategy name itself should reveal the origin of data.

Reply:
I meant group as in - any number of validators that should hold the full availability data.

Agree, i removed this field